proposals/extensions.md

```diff
@@ -50,6 +50,47 @@ receiver_parameter // add
+> - Additional local variable declaration spaces may occur within **extension declarations [and primary constructor parameter lists]. Names are introduced into these declaration spaces through member declarations.** Additional local variable declaration spaces may **also** occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *pattern*s, *declaration_expression*s,  *declaration_statement*s and *exception_specifier*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to contain elements with the same name. Thus, within a nested declaration space it is not possible to declare a local variable, local function or constant with the same name as a parameter, type parameter, local variable, local function or constant in an enclosing declaration space. It is possible for two declaration spaces to contain elements with the same name as long as neither declaration space contains the other. Local declaration spaces are created by the following constructs:
+>
+> ...
+
+Additions in **[square brackets]** are about the unrelated primary constructor feature, which is already in C# but not yet in the Standard. Extension declarations should behave the same way in terms of scoping.
+
+There is an existing need to specify (maybe in Section[12.8.4 Simple Names](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#1284-simple-names)) something along the lines of the following:
+
+> If a local variable directly occurring in a given declaration space is accessed from within a static member or static lambda expression that is itself nested within that declaration space, a compile-time error is given.
+
+Thus, for receiver parameters as well as any other local variable, a reference from within a static context still resolves to that variable, but leads to an error.
+
+``` c#
+public static class E
+{
+    extension(string s) // *1*
+    {
+        public int M(int i) // *2*
+        {
+            return s.Length + i;
+        }
+        public static string P => s; // Error: Cannot use s from static context
+    }
+}
+```
+
+In the example, `*1*` and `*2*` denote local variable declaration spaces, where `*2*` is nested within `*1*` and thus prohibited from introducing the same local variable names as `*1*`.
```

Comment:
Either I am misinterpreting what this statement is saying, or it looks like it contradicts the current language rules, which allow shadowing of both parameters and locals within nested methods. For example, the following code is legal:
```
class C2(int p)
{
    
    void M1(long p)
    {}

    static void M2(long p)
    {}

    void M3()
    {
        long p = 0;
    }

    static void M4()
    {
        long p = 0;
    }
    
    void M5(int p1)
    {
        string local1 (string p1) => p1;
        local1("");

        int l1 = 0;
        
        string local2 ()
        {
            string p1 = "";
            string l1 = "";
            return p1 + l1;
        }
        
        local2();
    }
}
```

Reply:
I think shadowing of locals and parameters within local functions and lambdas was allowed in Roslyn time frame. It is quite possible this relaxation didn't make it into the spec.

Reply:
Here is the relevant proposal - [https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/shadowing-in-nested-functions.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/shadowing-in-nested-functions.md). With primary constructor parameters we followed the same rules, i.e. they can be shadowed by member names and by parameters and locals declared by members.